### ISSUE TITLE
compass sprite-maps re-build on every save [6010]

### ISSUE THREAD

**Issue by hpowers** [bdkjones/CodeKit-1-Old#297]
The issue is with this function in compass:
http://compass-style.org/reference/compass/helpers/sprites/#sprite-map

Using it is causing 5+ minute delays saving any scss asset as all sprite-maps are re-built and re-calculated on every save even when nothing sprite related was modified.  SASS/COMPASS should see that the sprite maps are unchanged and leave them alone.

I did not have the issue in the last build I was using.  I am not sure what version that was, but Date Modified was Feb 20, 2012.  I am currently using 6010.  

I also fired up 'compass watch' on the directory and had no delays and sprite maps correctly did not re-build on every save, which would seem to confirm the issue is not with my markup.

**Comment by ghost** [comment]
I can confirm this issue. Everytime i make a minor change to a scss-file the sprite is regenerated. Saving/Reloading takes a while... :-/

Codekit 1.1.3 (6166) / included Compass Version: 0.12.1

**Comment by ghost** [comment]
I still have this issue in 1.14 (6204)... the sprites are not generated newly when using compass watch.

I create the sprites with:

```
$icon-spacing: 30px;
$icon-sprite-dimensions: true;
@import 'icon/*.png';
@include all-icon-sprites;
```

and then use the generated classes in HTML.

**Comment by ghost** [comment]
Still no luck with 1.2 (6449): on saving a .scss-file the sprite is generated every time. Every single time :-/
Do you need any more information on this?

**Comment by bdkjones** [comment]
Man, Compass is a pain in my ass. I'll have to dig in and take a look. Can one of you send me a project that exhibits the issue? I don't have any sprite projects with Compass.

**Comment by bdkjones** [comment]
Working with Chris Eppstein to track this one down.

**Comment by ghost** [comment]
Yeah.

**Comment by isoceles** [comment]
Any headway on this issue? I'm having the same problem. It takes Codekit 5+ minutes to refresh projects/framework every time I save, and believe it is due to the sprite issue stated above. Is there an option or a way to bypass the whole sprite generation when changes aren't made to the images?

**Comment by bdkjones** [comment]
Yes. You can add a line to your Compass config.rb file that bypasses the sprite generation. See the Compass docs for details, because I can't remember it off the top of my head.

**Comment by isoceles** [comment]
Okay, thanks!

**Comment by brendanfalkowski** [comment]
Any update on this? I didn't see any solution in the Compass thread linked by @piouPiouM above.

Whenever I edit an SCSS partial, Codekit regenerates all the sprites and the cooling fans start chugging. Frequently it errors out before compilation completes and restarts then finishes. This process takes roughly 30 seconds.

Or if I save a SCSS file again before compilation finishes, an error is logged and it restarts compilation. Then the fans go absolutely nuts.

The real barrier here is the pre-processor used to run faster than I could switch the browser to test. With sprites though it's really slowing down my workflow because any CSS change puts me at least 30 seconds behind.

**Comment by isoceles** [comment]
Unfortunately not. The guys at Compass said it's a CodeKit issue. (see below)

Scottdavis said:

> If you are using compass (0.12.2) you will need to upgrade for this to work properly. Compass does not regenerate a sprite unless its signature changes.
> 
> https://github.com/chriseppstein/compass/blob/stable/lib/compass/sass_extensions/sprites/sprite_methods.rb#L79-L81
> 
> If you are having this issue with Codekit it is probably related to the way they are calling compass and they should fix the issue on their end.

This issue was hampering my production so much I commented out my partial that contains all my sprites/images so I could continue to develop at a decent speed. I uncomment the partial when I need to show the team where I'm at.

**Comment by bdkjones** [comment]
I talked to Chris Eppstein about it. It _WAS_ an issue with Compass, but they have apparently accepted a patch in the past couple weeks to fix it. 12.2 is (as far as I can tell from the Compass GitHub repo) the currently-shipping version, however.

When Compass ships an update, I'll push that to CodeKit.

**Comment by 4rn0** [comment]
This is still an issue for me. Has the Compass update been shipped in Codekit yet? Or is it just me?

**Comment by brendanfalkowski** [comment]
I'm still having the issue also: https://twitter.com/falkowski/status/337205675564347392

It's actually faster for me to not use CodeKit and just push to my server for compilation than to work locally. I'd love to not have to do that, and would gladly buy another license if this was fixed.

**Comment by bdkjones** [comment]
Nope. Compass is still at 12.2.

I talked to @chriseppstein and verified that CodeKit is not using the --force flag when it tells Compass to compile a file (which would trigger a complete re-build of the sprites). But the issue persists because of a bug in Compass itself. There's a line that erroneously adds the --force flag to the command in certain circumstances. One of my users tracked it down and opened an issue on the Compass GitHub page.

If you're willing to do a bit of hacking, you can right-click the CodeKit app icon, choose "show package contents" and then find the resources/engines/Compass folder. In one of Compass's files, you change a single line of code and the issue disappears. The full details are on the appropriate issue over on Compass's GitHub page.

I can't guarantee that this change is safe, but it works for the guy that found it. It may be a good workaround until the Compass guys get around to releasing an update.

**Comment by bdkjones** [comment]
I have to issue an update to CodeKit later today to deal with some Jade nonsense and given that Compass hasn't released an update in nearly a year, I may just go ahead and add the hack to 12.2 so that this issue is solved. Check back for an update later tonight (let's say, 10PM PDT).

**Comment by 4rn0** [comment]
That sounds just lovely!
Thanks in advance!

:-)

**Comment by bdkjones** [comment]
FYI, here's the line that needs to be killed in Compass, in case you're interested: https://github.com/michaelgilley/compass/commit/f87c47d7d0886bd75cb0ad58194499fee28533dd

**Comment by 4rn0** [comment]
Any word on that update? :-)

**Comment by bdkjones** [comment]
Was waiting on a CoffeeScript update to fix a bug there. They pushed it yesterday, so I'll get to the CK update in the next couple days.

**Comment by 4rn0** [comment]
Thanks!

**Comment by clementoriol** [comment]
Hi !
I'm still having this issue, with CodeKit / Compass rebuilding sprites at every save.
Is it still supposed to happen ?

Here's how I'm generating my sprites : 

```
.folder-mySprite {
  height: folder-sprite-height(mySprite);
  width: folder-sprite-width(mySprite);
  @media (-webkit-min-device-pixel-ratio: 1.5), (-o-min-device-pixel-ratio: 3/2), (min-device-pixel-ratio: 1.5) {
      background-position: 0 round(nth(sprite-position($sprites_2x, 'mySprite'), 2) / 2);
  }
}
```

**Comment by brendanfalkowski** [comment]
It's working for me — only recompiling when an asset changes. Double-check CodeKit is updated?

**Comment by clementoriol** [comment]
Damn. And yes, Codekit is updated, bought it a few days ago for continuing on a project someone made using it.
Any ideas where it could come from ? I tried rebuilding the config.rb file without any luck.

**Comment by brendanfalkowski** [comment]
My sprite mixin is based on this: http://workofjonas.com/blog/2012/8/30/responsive-image-strategy
